src/aspire/source/image.py

```diff
@@ -339,36 +474,69 @@ def set_metadata(self, metadata_fields, values, indices=None):
             values should either be a scalar or a vector of length equal to the total number of images, |self.n|.
         :return: On return, the metadata associated with the specified indices has been modified.
         """
-        # Convert a single metadata field into a list of single metadata field, since that's what the 'columns'
-        # argument of a DataFrame constructor expects.
+
+        # This breaks lots of things, maybe not something we want to rush out.
+        # # Check if we're in an immutable state.
+        # if not self._mutable:
+        #     raise RuntimeError("This source is no longer mutable, try using `update` instead of `set_metadata`")
+
         if isinstance(metadata_fields, str):
             metadata_fields = [metadata_fields]
 
         if indices is None:
-            indices = self._metadata.index.values
+            indices = np.arange(self.n)
 
-        df = pd.DataFrame(values, columns=metadata_fields, index=indices)
-        for metadata_field in metadata_fields:
-            series = df[metadata_field]
-            if metadata_field not in self._metadata.columns:
-                self._metadata = self._metadata.merge(
-                    series, how="left", left_index=True, right_index=True
+        try:
```

Comment:
Describe briefly what's going on here? Some sort of iterable check, otherwise we broadcast?

Reply:
Added too comments

```
+        # Check if we're an iterable, and in case we're not, broadcast
+        # the single values into a list `indices` long.
         try:
             iter(values)
         except TypeError:
             values = [values] * len(indices)
         else:
+            # Special case for single `str`, which are iterable, but
+            # need to be broadcast like a singleton.
             if isinstance(values, str):
                 values = [values] * len(indices)
```

Reply:
Ok great. Thanks.